c++/src/kj/exception.c++

```diff
@@ -1070,9 +1076,30 @@ private:
 };
 
 ExceptionCallback& getExceptionCallback() {
-  static ExceptionCallback::RootExceptionCallback defaultCallback;
+  static auto defaultCallback = new ExceptionCallback::RootExceptionCallback();
+  // We allocate on the heap because some objects may throw in their destructors. If those objects
+  // had static storage, they might get fully constructed before the root callback. If they however
+  // then throw an exception during destruction, there would be a lifetime issue because their
+  // destructor would end up getting registered after the root callback's destructor. One solution
+  // is to just leak this pointer & allocate on first-use. The cost is that the initialization is
+  // mildly more expensive (+ we need to annotate sanitizers to ignore the problem). A great
+  // compiler annotation that would simply things would be one that allowed static variables to have
+  // their destruction omitted wholesale. That would allow us to avoid the heap but still have the
+  // same robust safety semantics leaking would give us. A practical alternative that could be
+  // implemented without new compilers would be to define another static root callback in
+  // RootExceptionCallback's destructor (+ a separate pointer to share its value with this
+  // function). Since this would end up getting constructed during exit unwind, it would have the
+  // nice property of effectively being guaranteed to be evicted last.
+  //
+  // All this being said, I came back to leaking the object is the easiest tweak here:
+  //  * Can't go wrong
+  //  * Easy to maintain
+  //  * Throwing exceptions is bound to do be expensive and malloc-happy anyway, so the incremental
+  //    cost of 1 heap allocation is minimal.
+  __lsan_ignore_object(defaultCallback);
```

Comment:
Well, that seems more correct. What if LSAN is putting these pointers into a list somewhere, and each time you call __lsan_ignore_object() you're actually appending to that list? Then you've created a new kind of leak, since every time we call getExceptionCallback() -- which is quite often -- it's extending the list.

I don't know that it's implemented like that, but I also don't know that it isn't. Writing this code the right way seems easy, can we just do it? I'd like to have a habit of doing things right when it's easy to do so, rather than rely on assumptions that may or may not be accurate.